web-console/src/components/show-log/show-log.tsx

```diff
@@ -43,70 +45,86 @@ export interface ShowLogProps {
 }
 
 export interface ShowLogState {
-  logValue: string;
+  logValue?: string;
+  loading: boolean;
+  error?: string;
   tail: boolean;
 }
 
 export class ShowLog extends React.PureComponent<ShowLogProps, ShowLogState> {
+  private showLogQueryManager: QueryManager<null, string>;
+  private tailQueryManager: QueryManager<null, void>;
   public log = React.createRef<HTMLTextAreaElement>();
 
   constructor(props: ShowLogProps, context: any) {
     super(props, context);
     this.state = {
-      logValue: '',
       tail: true,
+      loading: true,
     };
-    this.getLogInfo();
+
+    this.showLogQueryManager = new QueryManager({
+      processQuery: async () => {
+        const { endpoint, tailOffset } = this.props;
+        const resp = await axios.get(endpoint + (tailOffset ? `?offset=-${tailOffset}` : ''));
+        const data = resp.data;
+
+        let logValue = typeof data === 'string' ? data : JSON.stringify(data, undefined, 2);
+        if (tailOffset) logValue = removeFirstPartialLine(logValue);
+        return logValue;
+      },
+      onStateChange: ({ result, loading, error }) => {
+        this.setState({
+          logValue: result,
+          loading,
+          error,
+        });
+      },
+    });
+
+    this.tailQueryManager = new QueryManager({
+      processQuery: async () => {
+        await this.showLogQueryManager.runQuery(null);
+        if (this.state.tail) {
+          if (this.log.current) {
+            this.log.current.scrollTo(0, this.log.current.scrollHeight);
+          }
+          timeOut = Number(
+            setTimeout(() => {
+              this.tailQueryManager.runQuery(null);
+            }, 2000),
```

Comment:
`processQuery` should be a pure function with no side-effects (like setting a timeout), use set interval instead

Reply:
also set/clear the interval in component lifecycle methods (did mount / will unmount)